### ISSUE TITLE
Support for different types of bindings

### ISSUE THREAD

**Issue by ekinbarut** [CoreWCF/CoreWCF#429]
I couldn't find it on the roadmap if/when the other types of bindings in WCF will be ported to WCF Core and if there is a prioritization involved around them.

Missing types of bindings: 
- WSDualHttpBinding 
- WebHttpBinding
- NetPeerTcpBinding
- NetNamedPipeBinding
- NetMSMQBinding

**Comment by birojnayak** [comment]
@ekinbarut here is for NamedPipe => https://github.com/CoreWCF/CoreWCF/issues/234#issuecomment-703429293 and MSMQ => https://github.com/CoreWCF/CoreWCF/issues/234#issuecomment-703429857 ... I will get other bindings to the roadmap items soon..

**Comment by mconnew** [comment]
@ekinbarut, there are various challenges to each of those. I'm going to summarize (ie brain dump) what they are for each of them.

WSDualHttpBinding is a tough one as it mixes client and server. Basically a client creates an internal soap service to receive replys. Normally with HTTP the ReplyTo soap header is the anonymous header, which basically means send the reply over the natural transport reply mechanism. So that means the reply gets sent to the client in the HTTP response body. With WSDualHttpBinding, the client has it's own soap service created on it's own URL and sets that as the ReplyTo address in the soap header. So when you send a request, you get a 201 accepted HTTP response, and when the reply is ready, the server creates a SOAP client to send the response to the ReplyTo address which is received by the embedded soap service in the client.  
We could do this, but CoreWCF would need to own the client and the server packages as the WCF client project can't take dependency on CoreWCF due to not being officially Microsoft owned and supported. It's quite a bit of work as we haven't built out support yet for IInputChannel and IOutputChannel or CompositeDuplexBinding. As the whole channel layer has been reworked, I'm not sure how much work is left to support this. If possible, I would recommend investigating if you can switch to using WebSockets as we have support for that today.  

WebHttpBinding, this is a lower priority as you can use ASP.NET Core WebApi to achieve the same thing and we're focusing on scenarios where there isn't any other option. If you or someone else wants to volunteer to port it, I can absolutely support you in doing so and give you guidance on where to start. There's some architectural changes that would need to be done as it should probably use it's own transport implementation instead of using HttpTransportBindingElement, but honestly that just makes things easier as today it's doing a lot of shoehorning and making SOAP opinionated code do non-soap things.  

NetPeerTcpBinding is a client only binding which also fires up a soap service internally. I've spent a lot of time thinking about this over the years. This needs to be abstracted as the PNRP capability in Windows has been deprecated for a number of years. I want to replace it with the combo of mDNS (multicast DNS allows local network decentralized DNS record lookups) and DNS-SD (dns service discovery, basically record definitions to register and locate services via DNS). These two together would allow a mixed environment interop, and would also provide a forward path for those using NetPeerTcpBinding today which has been deprecated and on borrowed time. I'd want to make it pluggable for service discovery so we could still provide the PNRP based mechanism for interop on WIndows, and also bridge between the two systems (NetPeerTcpBinding provides for a mesh like mechanism for advertising to peers potentially unknown hosts so can bridge between 2 different peer discovery mechanisms). This is one of those projects which I would have a lot of fun creating, but is going to take a while to get to due to low usage reducing the priority.  

NetNamedPipeBinding is also a tricky one. It requires a custom transport as ASP.NET Core doesn't support NamedPipes. It does support Unix Domain Sockets though which provides a lot of the same capabilities. I have been informed by someone in the ASP.NET Core team that UDS is faster than localhost sockets on Windows, and as localhost sockets are faster than named pipes on Windows, UDS is something which is worth supporting on Windows. But we do also need NamedPipe support for interop with NetFx. It's also going to be needed to communicate with TCP port sharing on Windows. To create a custom transport alongside Kestrel will mean requiring a later version of ASP.NET Core, which is why it can't happen until we drop support for ASP.NET Core 2.1. WSDL support needs to happen before that. So I need to create support for NamedPipes, and UDS. They share a lot of implementation with NetTcp (about 90% common code), so some refactoring needs to happen to make a general purpose NetFraming transport which NamedPipes, UDS and NetTcp all depend on and extend to provide their 10% of logic. So there's a non-trivial amount of work to port it. The good news though is that I had this in mind when I re-wrote the NetTcp binding so it's already been restructured to make this easier.  

NetMSMQBinding is an interesting one. It's bound to Windows only because of the MSMQ dependency. I want to create a more generic Queue based transport binding which is extendable, and provide the implementation for MSMQ. This should make it easy for others to contribute implementations for other queue based systems such as AMQP as all you would need to do is have a mechanism to retrieve a queued item and return a byte array of the body of the queue item. There's some complication around how to configure dead letter queues, and timeouts around messages being retrieved but not finished processing (i.e. service is restarted while processing so some items needs to be reprocessed), but I suspect that will become clear on how to handle once MSMQ is prototyped.

**Comment by Ximik87** [comment]
@mconnew I have a little experience with queues (based on working with NServiceBus framework) and I can try to make prototypes of bindings for working with MSMQ and RabbitMQ

**Comment by mconnew** [comment]
@Ximik87, let's take this discussion to #164 which is the issue being used to track MSMQ/generic queuing support. I'll write up a high level design of how this should be implemented and we can go from there. It will also get visibility from people who are interested in this so might get some useful feedback about the design discussion.
